src/bms/model/control_flow.rs

```diff
@@ -0,0 +1,650 @@
+    where
+        T: IntoIterator<Item = Token<'a>>,
+    {
+        let v = tokens
+            .into_iter()
+            .map(NonControlToken::try_from_token)
+            .flat_map(Result::ok)
+            .collect();
+        Self::Tokens(v)
+    }
+
+    /// Convert this unit into lex tokens.
+    #[must_use]
+    pub fn into_tokens(self) -> Vec<Token<'a>> {
+        match self {
+            TokenUnit::Random(r) => r.into_tokens(),
+            TokenUnit::Switch(s) => s.into_tokens(),
+            TokenUnit::Tokens(v) => v.into_iter().map(Token::from).collect(),
+        }
+    }
+}
+
+impl<'a> From<Random<'a>> for TokenUnit<'a> {
+    fn from(value: Random<'a>) -> Self {
+        TokenUnit::Random(value)
+    }
+}
+
+impl<'a> From<Switch<'a>> for TokenUnit<'a> {
+    fn from(value: Switch<'a>) -> Self {
+        TokenUnit::Switch(value)
+    }
+}
+
+impl<'a> From<Vec<NonControlToken<'a>>> for TokenUnit<'a> {
+    fn from(value: Vec<NonControlToken<'a>>) -> Self {
+        TokenUnit::Tokens(value)
+    }
+}
+
+/// Indicates whether the random block generates a value (`#RANDOM`) or uses a set value (`#SETRANDOM`).
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ControlFlowValue {
+    /// Use a fixed value (emits `#SETRANDOM <value>`).
+    Set(BigUint),
+    /// Generate a random value with the given maximum (emits `#RANDOM <max>`).
+    GenMax(BigUint),
+}
+
+/// One branch node in an if-chain inside a random block.
+#[derive(Debug, Clone, PartialEq, Eq)]
+enum IfChainEntry<'a> {
+    /// An `#ELSEIF <cond>` branch with its units and a pointer to the next entry.
+    ElseIf {
+        /// Condition value for `#ELSEIF <cond>`.
+        cond: BigUint,
+        /// Content units for this branch.
+        units: Vec<TokenUnit<'a>>,
+        /// Pointer to the next chain entry.
+        next: Box<IfChainEntry<'a>>,
+    },
+    /// An `#ELSE` branch with its units. This is always terminal.
+    Else {
+        /// Content units for this branch.
+        units: Vec<TokenUnit<'a>>,
+    },
+    /// Terminator for an if-chain.
+    EndIf,
+}
+
+impl<'a> IfChainEntry<'a> {
+    fn append_to_tail(&mut self, entry: Self) {
+        match self {
+            IfChainEntry::EndIf => {
+                *self = entry;
+            }
+            IfChainEntry::ElseIf { next, .. } => {
+                let mut cur = next.as_mut();
+                loop {
+                    match cur {
+                        IfChainEntry::EndIf => {
+                            *cur = entry;
+                            break;
+                        }
+                        IfChainEntry::ElseIf { next, .. } => {
+                            cur = next.as_mut();
+                        }
+                        IfChainEntry::Else { .. } => {
+                            // ELSE is terminal; do not append beyond it.
+                            break;
+                        }
+                    }
+                }
```

Comment:
It can be done by a recursion.
```suggestion
                next.append_to_tail(entry);
```